xray/src/client.rs

```diff
@@ -0,0 +1,231 @@
+        tokio::pin!(handshake_timeout);
+        loop {
+            tokio::select! {
+                Ok(recv_type) = self.recv_encrypted(&mut handshake_buf) => {
+                    if matches!(recv_type, RecvType::Handshake) {
+                        println!("Handshake: done");
+                        return Ok(());
+                    }
+                }
+                _ = &mut handshake_timeout => {
+                    return Err(XRayError::HandshakeTimedOut);
+                }
+            }
+        }
+    }
+
+    pub async fn send_packet(
+        &mut self,
+        sock_dst: SocketAddrV4,
+        packet_dst: SocketAddrV4,
+        payload: &[u8],
+    ) -> XRayResult<SendType> {
+        if self.tunn.is_some() {
+            self.send_encrypted(sock_dst, packet_dst, payload).await
+        } else {
+            self.send_plaintext(sock_dst, payload).await
+        }
+    }
+
+    pub async fn recv_packet(&mut self, buf: &mut [u8]) -> XRayResult<RecvType> {
+        if self.tunn.is_some() {
+            self.recv_encrypted(buf).await
+        } else {
+            self.recv_plaintext(buf).await
+        }
+    }
+
+    pub async fn tick_timers(&mut self, wg_addr: SocketAddrV4) {
+        if let Some(ref mut tunn) = &mut self.tunn {
+            match tunn.update_timers(&mut self.buf) {
+                TunnResult::Done => (),
+                TunnResult::WriteToNetwork(packet) => {
+                    if let Err(e) = self.sock.send_to(packet, wg_addr).await {
+                        println!("Update Tunn timers failed to send to socket: {:?}", e);
+                    }
+                }
+                TunnResult::Err(e) => {
+                    println!("Failed to update Tunn timers: {:?}", e);
+                }
+                unexpected_result => println!(
+                    "Update Tunn timers returned unexpected result: {:?}",
+                    unexpected_result
+                ),
+            }
+        }
+    }
+
+    async fn send_encrypted(
+        &mut self,
+        sock_dst: SocketAddrV4,
+        packet_dst: SocketAddrV4,
+        payload: &[u8],
+    ) -> XRayResult<SendType> {
+        let tunn = &mut self
+            .tunn
+            .as_mut()
+            .expect("This function should only be called on clients with a Tunn object");
+        let packet = Self::make_udp_packet(self.addr, packet_dst, payload)?;
+        let tr = tunn.encapsulate(&packet, &mut self.buf);
+        let (send_type, packet): (SendType, &[u8]) = match tr {
+            TunnResult::Done => (SendType::None, &[]),
+            TunnResult::WriteToNetwork(p) => match p[0] {
+                1 => (SendType::Handshake, p),
+                4 => (SendType::Data, p),
+                unexpected => return Err(XRayError::UnexpectedPacketType(unexpected)),
+            },
+            TunnResult::WriteToTunnelV4(p, _) => (SendType::Tunn, p),
```

Comment:
We should not recieve this (basically it' unexpected), and also, if we do, it should not go into crypto socket, as it's a plaintext packet, intended fot the tunnel

Reply:
Good point. That case is now considered `unexpected`